### PR TITLE
Architecture rework

### LINKED ISSUE #32: Implement generic persistent data storage
**Is your feature request related to a problem? Please describe.**
The current sqflite storage takes about 100 lines of code for a class with a foreign key (see the [implementation of `ArticleDao`](https://github.com/schul-cloud/schulcloud-flutter/blob/master/lib/news/data/repository.dart)).

**Describe the solution you'd like**
Any solution where we wouldn't have to code the mapping to the database object by hand would be better suited. Perhaps [hive](pub.dev/packages/hive) might be worth a shot.
Either way, the database solution could be combined with a custom annotation `@Entity` that generates the database adapter-specific stuff for us.

### LINKED ISSUE #41: Put ApiService's logic into BLoCs
**Is your feature request related to a problem? Please describe.**
With implementing all the APIs, the ApiService becomes more and more monolithic.

**Describe the solution you'd like**
I think it would be best to remove the ApiService and put its logic directly into the BLoCs instead.

### LINKED ISSUE #42: Remove core module
**Is your feature request related to a problem? Please describe.**
The git repo of the app should really only contain app-specific code.

**Describe the solution you'd like**
Meta-concepts like Repositories that can be used in a variety of apps should rather be defined in separate git repos and then be pub-lished.

### PR DESCRIPTION
**Pull request by MarcelGarus**
Fixes #32
Fixes #41
Fixes #42
Contrary to the name of this branch, the logout is not implemented yet, sorry about that. 👀

**So, what exactly is new?**
* The core module is gone! Per definition, it only contained code that's not app-specific. This has been outsourced into public dart packages named [`repository`](https://pub.dev/packages/repository), [`repository_hive`](https://pub.dev/packages/repository_hive) and [`repository_shared_preferences`](https://pub.dev/packages/repository_shared_preferences).
* Speaking of `repository_hive`, we now use [`hive`](https://pub.dev/packages/hive) for storing data locally. This is not used extensively throughout the app yet (only in the news and app module, if I remember correctly), but it's very easy to integrate custom types.
* The file structure inside the modules also changed a little bit—there's no `data` directory anymore, rather there's just a `data.dart` file. Extensive research showed that we only used about 1-3 data classes in the modules anyway, so merging those into one file leaves us with less boilerplate.
* The services got fancier! There's no more `ApiService` anymore, rather the module-specific logic went into the modules themselves. So, no more wrapper repository just to call `api.listNews()`, it happens right there. Also, a new utility repository `CollectionDownloader` is now in the `app/utils.dart`. You just need to implement one simple method for downloading all the items—no more manual maintaining of a `_downloader` and checking `if (_items == null) await _downloader;` in the fetch methods!
* The service initialization also got updated. The `SplashScreen` now waits for all services to be initialized and only then redirects to the actual app. Very fancy!
* The imports got a lot cleaner! There were some interdependencies between modules, which isn't necessarily bad, but made it harder to reason about the effect of changes when working on a module. Now, anything used outside of a module is exported in the `<module>/<module>.dart`, so other packages only need to import the top-level module file, i.e. `app/app.dart` instead of `app/utils.dart`, `app/widgets.dart` etc.

**Known problems?**
* The files screen no longer contains the courses. The reason is there's a strong linkage between the `courses` and the `files` modules—because the file screen also displays all the courses, it has logic quite similar to the course screen. Perhaps, we can fix this by implementing a new `CourseService` of sorts, that allows for downloading the courses and is used among the modules.

### CODE REVIEW

File: lib/app/widgets/menu.dart [hunk 1]
```diff
@@ -17,6 +17,11 @@ class Menu extends StatefulWidget {
 class _MenuState extends State<Menu> {
   void _navigateTo(Routes target) => Navigator.pop(context, target.name);
 
+  Future<void> _logOut() async {
+    await Provider.of<AuthenticationStorageService>(context).logOut();
+    Navigator.of(context).pushReplacementNamed(Routes.login.toString());
```
Comment (ctiedt) [inline lib/app/widgets/menu.dart]: `Routes.login.toString()` should be `Routes.login.name` as the former will confuse the navigator.

File: lib/courses/bloc.dart [hunk 2]
```diff
@@ -1,42 +1,147 @@
...
-
-  Stream<List<Lesson>> getLessons(Id<Course> courseId) {
-    var _lessons = CachedRepository<Lesson>(
-      source: LessonDownloader(
-        api: api,
-        courseId: courseId,
-      ),
+  Stream<List<Lesson>> getLessonsOfCourse(Id<Course> courseId) {
+    return CachedRepository<Lesson>(
+      source: LessonDownloader(network: network, courseId: courseId),
```
Comment (ctiedt) [inline lib/courses/bloc.dart]: As this reloads the lessons every time it is called, the cache seems superfluous here.

File: lib/app/services/me.dart [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    if (token == null)
+      _meSubject.add(null);
+    else {
+      final id = Id<User>(_decodeTokenToUser(token));
+      final me = await user.fetchUser(id);
+      _meSubject.add(me);
+    }
+  }
+
+  // A JWT token exists of a header, body and claim (signature), all separated
```
Comment (ctiedt) [inline lib/app/services/me.dart]: "exists" should be changed to "consists"

File: lib/courses/bloc.dart [hunk 4]
```diff
@@ -1,42 +1,147 @@
...
+        id: Id<Course>(data['_id']),
+        name: data['name'],
+        description: data['description'],
+        teachers: await Future.wait([
+          for (String id in data['teacherIds']) user.fetchUser(Id<User>(id)),
+        ]),
+        color: hexStringToColor(data['color']),
+      );
+    }));
+    print('Courses done');
```
Comment (ctiedt) [inline lib/courses/bloc.dart]: This debug print statement should be removed.

File: lib/files/widgets/files_screen.dart [hunk 5]
```diff
@@ -50,28 +43,31 @@ class UserFilesCard extends StatelessWidget {
...
-              body: FilesView(owner: Provider.of<UserService>(context).userId),
-              bottomNavigationBar: MyAppBar(),
+      context,
+      MaterialPageRoute(
+        builder: (context) => ProxyProvider<NetworkService, FilesService>(
+          builder: (_, network, __) => FilesService(
+            network: network,
+            owner: Provider.of<MeService>(context).me.id.toString(),
+          ),
+          child: Scaffold(
```
Comment (ctiedt) [inline lib/files/widgets/files_screen.dart]: As the `FilesView` has its own `Scaffold` , it should not be wrapped in one here.

File: ios/Flutter/flutter_export_environment.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+# This is a generated file; do not edit or check into version control.
```
Comment (ctiedt) [inline ios/Flutter/flutter_export_environment.sh]: As the file says, it should not be included in the git repository.

File: .dart_tool/build/098d3ee73e6cc294616d3a2e2c3c81ad/asset_graph.json [hunk 7]
```diff
@@ -0,0 +1 @@
...51,152,153,154,155,156,157,158,159,160,161,162,164,165,166,167,168,169,170,171,172,176,196,207],[],"z4iP9cPa74fjcLaGmgfIXw==",[]],[0,1118,[77,71,75,80,81,72,67,73,74,70,65,69,78,66,79,76,83,84,64,85,88,87,89,86,90,91,92,94,96,97,95,102,105,103,104,107,109,110,111,100,101,98,99,106,112,113,114,115,116,117,119,120,121,122,124,125,126,127,128,129,130,131,132,133,134,135,136,138,139,140,141,142,143,144,145,146,147,149,150,151,152,153,154,155,156,157,158,159,160,161,162,164,165,166,167,168,169,170,171,172,176,196,207],[],"TqkJK8FevYvjtbQdKM0hGw==",[]],[0,1119,[77,71,75,80,81,72,67,73,74,70,65,69,78,66,79,76,83,84,64,85,88,87,89,86,90,91,92,94,96,97,95,102,105,103,104,107,109,110,111,100,101,98,99,106,112,113,114,115,116,117,119,120,121,122,124,125,126,127,128,129,130,131,132,133,134,135,136,138,139,140,141,142,143,144,145,146,147,149,150,151,152,153,154,155,156,157,158,159,160,161,162,164,165,166,167,168,169,170,171,172,176,196,207],[],"36RExbYljLK1i8/83Gw+kg==",[]],[0,1120,[77,71,75,80,81,72,67,73,74,70,65,69,78,66,79,76,83,84,64,85,88,87,89,86,90,91,92,94,96,97,95,102,105,103,104,107,109,110,111,100,101,98,99,106,112,113,114,115,116,117,119,120,121,122,124,125,126,127,128,129,130,131,
```
Comment (ctiedt) [inline .dart_tool/build/098d3ee73e6cc294616d3a2e2c3c81ad/asset_graph.json]: This appears to be an automatically generated file and should not be included in any commits.

File: .dart_tool/build/entrypoint/build.dart [hunk 8]
```diff
@@ -0,0 +1,29 @@
+// ignore_for_file: directives_ordering
```
Comment (ctiedt) [inline .dart_tool/build/entrypoint/build.dart]: This appears to be an automatically generated/build related file and should not be included in any commits.

**Review by ctiedt — CHANGES_REQUESTED** [review]
It appears that you included a number of automatically generated files (in particular the `.dart_tool` and `.idea` directories) that should normally be excluded  through your `.gitignore`.

File: lib/app/widgets/menu.dart (hunk 1 above)
Comment (MarcelGarus) [reply lib/app/widgets/menu.dart]: Because the logout doesn't work anyway, I commented it out for now.

**Review by ctiedt — APPROVED** [review]

**Pull request merged**